lib/pbench/agent/tool_meister_client.py

```diff
@@ -41,195 +34,173 @@ class Client:
+        cls,
+        signal_publisher: state_signals.SignalExporter,
+        logger: logging.Logger = None,
+    ):
+        """
+        Creates the Client object using an existing, passed-in
+        SignalExporter.
+        """
 
-    def publish(self, group, directory, action, args=None):
-        """publish - marshal a JSON document formed from the group, directory,
-        action, and args arguments.
+        return cls(signal_publisher, False, logger)
+
+    @classmethod
+    def create_with_redis(
+        cls,
+        existing_redis_client: redis.Redis = None,
+        redis_host: str = "localhost",
+        redis_port: str = "6379",
+        publisher_name: str = "pbench_client",
+        logger: logging.Logger = None,
+    ):
+        """
+        Creates the Client object with a new SignalExporter,
+        built either with an existing redis connection object
+        or with a host+port specification.
+        """
+
+        if existing_redis_client:
+            sig_pub = state_signals.SignalExporter(
+                publisher_name, existing_redis_conn=existing_redis_client
+            )
+        else:
+            sig_pub = state_signals.SignalExporter(
+                publisher_name, redis_host=redis_host, redis_port=redis_port
+            )
+        sig_pub.initialize_and_wait(
+            1,
+            list(tm_allowed_actions),
+            tag="from_pbench_client",
+        )
+        return cls(sig_pub, True, logger)
+
+    def _publish(self, group: str, directory: str, action: str, args=None) -> int:
+        """
+        A helper function that handles the publishing of messages via
+        state-signals, using the Client's SignalExporter. It then awaits
+        responses from subscribers (TDS).
 
         Returns 0 on success, 1 on failure; logs are also written for any
         errors encountered.
         """
-        if action not in api_tm_allowed_actions:
-            return 1
+
         # The published message contains four pieces of information:
         #   {
-        #     "action": "< 'start' | 'stop' | 'send' | 'kill' >",
+        #     "action": "< 'init' | 'start' | 'stop' | 'send' | 'end' | 'terminate' >",
         #     "group": "< the tool group name for the tools to operate on >",
         #     "directory": "< the local directory path to store collected data >"
         #     "args": "< arbitrary argument payload for a particular action >"
         #   }
         # The caller of tool-meister-client must be sure the directory argument
         # is accessible by the Tool Data Sink instance.
-        self.logger.debug("publish %s on chan %s", action, self.from_client_channel)
-        msg = dict(action=action, group=group, directory=str(directory), args=args)
+
+        # This check/conversion is necessary, as the pbench scripts sometimes pass
+        # in the directory as a str, and sometimes as pathlib.PosixPath
+        if directory and not isinstance(directory, str):
```

Comment:
Why do we have to make this check now?  What changed that is causing this problem?

Reply:
I've just formalized/cleaned-up to old "cast to string" line (before it was just `if directory: directory=str(directory)`. This is done because some pbench scripts pass a string path into the Client object, and others pass in a PosixPath object, which would result in the following error if left unconverted: `TypeError: Object of type PosixPath is not JSON serializable`

Reply:
Didn't the original code already handle that with `str(directory)` without the test?

Reply:
In cases where `directory=None`, simply doing `str(directory)` will convert `None` into a string of value `"None"`. Additionally I figured we should be gracefully catching the scenario where something that can't be cast to string is passed in.